engine/src/lhs_types/bytes.rs

```diff
@@ -0,0 +1,346 @@
+use serde::de::Visitor;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::borrow::Cow;
+use std::hash::{Hash, Hasher};
+use std::ops::Deref;
+
+/// A byte string.
+#[derive(Debug, Clone)]
+pub enum Bytes<'a> {
+    /// Borrowed byte string.
+    Borrowed(&'a [u8]),
+    /// Owned byte string.
+    Owned(Box<[u8]>),
+}
+
+impl<'a> Bytes<'a> {
+    /// Clones self into a fully owned byte string.
+    #[inline]
+    pub fn to_owned(&self) -> Bytes<'static> {
+        match self {
+            Self::Borrowed(b) => Bytes::Owned(Box::from(*b)),
+            Self::Owned(b) => Bytes::Owned(b.clone()),
+        }
+    }
+
+    /// Converts self into a fully owned byte string.
+    #[inline]
+    pub fn into_owned(self) -> Bytes<'static> {
+        match self {
+            Self::Borrowed(b) => Bytes::Owned(Box::from(b)),
+            Self::Owned(b) => Bytes::Owned(b),
+        }
+    }
+
+    /// Converts self into a `Vec<u8>` by re-using allocated memory when possible.
+    #[inline]
+    pub fn to_mut(&mut self) -> &mut [u8] {
+        if let Self::Borrowed(b) = self {
+            *self = Self::Owned(Box::from(*b));
+        }
+        match self {
+            Self::Owned(b) => b,
+            Self::Borrowed(_) => {
+                cfg_if::cfg_if! {
+                    if #[cfg(debug_assertions)] {
+                        unreachable!()
+                    } else {
+                        std::hint::unreachable_unchecked()
+                    }
+                }
+            }
+        }
+    }
+
+    /// Shortens the byte string, keeping only the first `len` elements.
+    #[inline]
+    pub fn truncate(&mut self, len: usize) {
+        match self {
+            Self::Borrowed(slice) => {
+                *slice = &slice[..len];
+            }
+            Self::Owned(data) => {
+                let mut vec = Vec::from(std::mem::take(data));
+                vec.truncate(len);
+                *data = Box::from(vec);
+            }
+        }
+    }
+}
+
+impl Deref for Bytes<'_> {
+    type Target = [u8];
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        match self {
+            Self::Borrowed(b) => b,
+            Self::Owned(b) => b,
+        }
+    }
+}
+
+impl AsRef<[u8]> for Bytes<'_> {
+    #[inline]
+    fn as_ref(&self) -> &[u8] {
+        match self {
+            Self::Borrowed(b) => b,
+            Self::Owned(b) => b,
+        }
+    }
+}
+
+impl<'a> From<&'a [u8]> for Bytes<'a> {
+    #[inline]
+    fn from(value: &'a [u8]) -> Self {
+        Bytes::Borrowed(value)
+    }
+}
+
+impl<'a, const N: usize> From<&'a [u8; N]> for Bytes<'a> {
+    #[inline]
+    fn from(value: &'a [u8; N]) -> Self {
+        Bytes::Borrowed(value)
+    }
+}
```

Comment:
nit: Maybe implement `From<[u8; N]>` too? Since we implement `PartialEq` against this anyway